File: leetcode/problems/2.go
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
  p := &ListNode{}
  head, last := p, 0
  for l1 != nil || l2 != nil {
    n1, n2 := 0, 0
    if l1 != nil {
      n1 += l1.Val
      l1 = l1.Next
    }
    if l2 != nil {
      n2 += l2.Val
      l2 = l2.Next
    }
    sum := n1 + n2 + last
    sum, last = sum % 10, sum / 10
    cur := &ListNode{Val: sum}
    p.Next = cur
    p = cur
  }
  if last != 0 {
    p.Next = &ListNode{Val: last}
  }
  return head.Next
}

File: leetcode/problems/2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode newNode=ListNode(0);
        ListNode *nowNode=&newNode;
        int rest=0;
        while(l1||l2||rest)
        {
            rest+=(l1?l1->val:0);
            rest+=(l2?l2->val:0);
            l1=(l1?l1->next:l1);
            l2=(l2?l2->next:l2);
            int val=rest%10;
            rest/=10;
            nowNode->next=new ListNode(val);
            nowNode=nowNode->next;
        }
        return newNode.next;
    }
};